op-node/l1/batching.go

```diff
@@ -2,134 +2,103 @@ package l1
-			requests = requests[len(nextBatch):]
-
-			// schedule the batch, this may block if all workers are busy and the queue is full
-			batchRequests <- nextBatch
+// Fetch fetches more of the data, and returns io.EOF when all data has been fetched.
+// This method is safe to call concurrently: it will parallelize the fetching work.
+// If no work is available, but the fetching is not done yet,
+// then Fetch will block until the next thing can be fetched, or until the context expires.
+func (ibc *IterativeBatchCall) Fetch(ctx context.Context, maxBatchSize uint) error {
+	// collect a batch from the requests channel
+	batch := make([]rpc.BatchElem, 0, maxBatchSize)
+	// wait for first element
+	select {
+	case reqElem, ok := <-ibc.scheduled:
+		if !ok { // no more requests to do
+			return io.EOF
 		}
-		return i
+		batch = append(batch, reqElem)
+	case <-ctx.Done():
+		return ctx.Err()
 	}
 
-	maxCount := expectedBatches * maxRetry
-
-	awaited := len(requests)
-
-	// start initial round of parallel requests
-	count := parallelRequests()
-
-	// We slow down additional batch requests to not spam the server.
-	retryTicker := time.NewTicker(time.Millisecond * 20)
-	defer retryTicker.Stop()
-
-	// The main requests slice is only ever mutated by the go-routine running this loop.
-	// Slices of this are sent to worker go-routines, and never overwritten with different requests.
+	// collect more elements, if there are any.
 	for {
-		// check if we've all results back successfully
-		if awaited <= 0 {
-			return nil
-		}
-		if count > maxCount {
-			return TooManyRetries
+		if uint(len(batch)) >= maxBatchSize {
+			break
 		}
 		select {
-		case <-retryTicker.C:
-			count += parallelRequests() // retry batch-requests on interval
-		case result := <-completed:
-			if result.err != nil {
-				// batch failed, RPC may be broken, abort
-				return fmt.Errorf("batch request failed: %w", result.err)
+		case reqElem, ok := <-ibc.scheduled:
+			if !ok { // no more requests to do
+				return io.EOF
 			}
-			// if any element failed, add it to the requests for re-attempt
-			requests = append(requests, result.failed...)
-			awaited -= result.success
+			batch = append(batch, reqElem)
+			continue
 		case <-ctx.Done():
 			return ctx.Err()
+		default:
+			break
 		}
+		break
 	}
+
+	if err := ibc.getBatch(ctx, batch); err != nil {
+		for _, r := range batch {
+			ibc.scheduled <- r
+		}
+		return fmt.Errorf("failed batch-retrieval: %w", err)
+	}
+	var result error
+	for _, elem := range batch {
+		if elem.Error != nil {
+			result = multierror.Append(result, elem.Error)
+			elem.Error = nil // reset, we'll try this element again
+			ibc.scheduled <- elem
+			continue
+		} else {
+			atomic.AddUint32(&ibc.completed, 1)
+			if ibc.completed >= uint32(len(ibc.requests)) {
+				close(ibc.scheduled)
+				return io.EOF
+			}
+		}
+	}
+	return result
+}
+
+func (ibc *IterativeBatchCall) Complete() bool {
+	return ibc.completed >= uint32(len(ibc.requests))
```

Comment:
Would it be a bug if `completed > ibc.requests`?

Reply:
Technically yes, but it doesn't hurt to do `>=` instead of `==`

Reply:
Perhaps a warning log could help give insights into this happening?